lib/HunkView.ml

```diff
@@ -67,91 +89,197 @@ let ui_hunk_summary (hunk : string Patch.hunk) : Nottui.ui =
+          (Printf.sprintf "%2d %2d   " (mine_num + 1) (their_num + 1))
 
-let current_hunks (z_patches : string Patch.t Zipper.t) : Nottui.ui =
-  let p = Zipper.get_focus z_patches in
-  let hunks = List.map ui_unified_diff p.Patch.hunks in
-  Ui.vcat hunks
+let render_word_diff (words : WordDiff.word list)
+    (diff_type : [ `Added | `Removed | `Unchanged ]) (mode : rendering_mode) :
+    Ui.t =
+  let render_word = function
+    | WordDiff.Changed word -> style_word word diff_type mode
+    | WordDiff.Unchanged word -> style_word word `Unchanged mode
+  in
+  Ui.hcat (List.map render_word words)
+
+let render_diff_line (mine_num : int) (their_num : int)
+    (diff_type : [ `Added | `Removed | `Unchanged ])
+    (content : WordDiff.word list) (mode : rendering_mode) : Ui.t =
+  let line_number = render_line_number mine_num their_num diff_type mode in
+  let content_ui = render_word_diff content diff_type mode in
+  Ui.hcat [ line_number; content_ui ]
 
-(** Side by side diff view implementation **)
+let render_hunk_lines (hunk_lines : WordDiff.line_content Patch.line list)
+    (mode : rendering_mode) : Ui.t =
+  let rec process_lines mine_num their_num acc = function
+    | [] -> List.rev acc
+    | line :: rest ->
+        let new_mine, new_their, ui =
+          match line with
+          | `Common words ->
+              ( mine_num + 1,
+                their_num + 1,
+                render_diff_line mine_num their_num `Unchanged words mode )
+          | `Mine words ->
+              ( mine_num + 1,
+                their_num,
+                render_diff_line mine_num their_num `Removed words mode )
+          | `Their words ->
+              ( mine_num,
+                their_num + 1,
+                render_diff_line mine_num their_num `Added words mode )
+        in
+        process_lines new_mine new_their (ui :: acc) rest
+  in
+  Ui.vcat (process_lines 0 0 [] hunk_lines)
 
-let lines_with_numbers (lines : line list) (attr_change : Notty.attr)
-    (prefix : string) : Nottui.ui list =
+let render_hunk (hunk : string Patch.hunk) (mode : rendering_mode) : Ui.t =
+  let content =
+    let blocks =
+      Block.of_hunk hunk.Patch.lines
+      |> List.map WordDiff.compute |> Block.to_hunk
+    in
+    render_hunk_lines blocks mode
+  in
+  Ui.vcat [ content ]
+
+(* Helper functions for side-by-side view *)
+
+let lines_with_numbers (lines : line list) (attr_change : Notty.attr option)
+    (change_type : [ `Add | `Remove | `Unchanged ]) (mode : rendering_mode) :
+    Ui.t list =
   let rec process_lines line_num acc = function
     | [] -> List.rev acc
     | line :: rest ->
-        let content, attr, next_num =
+        let line_ui, next_num =
           match line with
           | Common s ->
               let content = Printf.sprintf "%3d   %s" line_num s in
-              (content, Notty.A.empty, line_num + 1)
+              let attr =
+                match mode with
+                | Color -> Some Notty.A.(fg lightcyan)
```

Comment:
Turning common lines or words into light cyan is not related to the goal of this PR and should be reverted.